File: jjazz-pure-core/src/main/java/org/jjazz/purecore/harmony/Position.java
package org.jjazz.purecore.harmony;

import static com.google.common.base.Preconditions.checkArgument;
import java.util.Objects;

/**
 * A musical position defined by a bar index and a beat position within that bar.
 * Immutable. Extracted from JJazzLab Position.
 */
public final class Position implements Comparable<Position> {

    private final int bar;
    private final float beat;

    /** Position at bar 0, beat 0. */
    public Position() {
        this(0, 0f);
    }

    /** Position at given bar, beat 0. */
    public Position(int bar) {
        this(bar, 0f);
    }

    public Position(int bar, float beat) {
        checkArgument(bar >= 0, "bar=%s", bar);
        checkArgument(beat >= 0, "beat=%s", beat);
        this.bar = bar;
        this.beat = beat;
    }

    /** Copy constructor. */
    public Position(Position pos) {
        this(pos.bar, pos.beat);
    }

    public int getBar() { return bar; }

    /**
     * The beat position within the bar (0.0 = start of bar).
     */
    public float getBeat() { return beat; }

    /**
     * Integer part of the beat.
     */
    public int getBeatInt() { return (int) beat; }

    /**
     * Fractional part of the beat [0.0, 1.0).
     */
    public float getBeatFractionalPart() { return beat - (int) beat; }

    /**
     * Convert to absolute beat position in the given time signature.
     */
    public float toAbsoluteBeat(TimeSignature ts) {
        return bar * ts.getNbNaturalBeats() + beat;
    }

    /**
     * Create a Position from an absolute beat position and time signature.
     */
    public static Position fromAbsoluteBeat(float absoluteBeat, TimeSignature ts) {
        float nbBeatsPerBar = ts.getNbNaturalBeats();
        int bar = (int) (absoluteBeat / nbBeatsPerBar);
        float beat = absoluteBeat - bar * nbBeatsPerBar;
        return new Position(bar, beat);
    }

    @Override
    public int compareTo(Position o) {
        int cmp = Integer.compare(bar, o.bar);
        if (cmp == 0) cmp = Float.compare(beat, o.beat);
        return cmp;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Position p)) return false;
        return bar == p.bar && Float.compare(beat, p.beat) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(bar, beat);
    }

    @Override
    public String toString() {
        return "bar:" + bar + ":" + String.format("%.3f", beat);
    }
}

File: jjazz-engine/src/harmony/position.rs
use crate::harmony::time_signature::TimeSignature;

/// A musical position defined by bar index and beat within that bar.
#[derive(Debug, Clone, Copy, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct Position {
    pub bar: i32,
    pub beat: f32,
}

impl Position {
    pub fn new(bar: i32, beat: f32) -> Self {
        assert!(bar >= 0 && beat >= 0.0);
        Self { bar, beat }
    }

    pub fn bar_zero() -> Self { Self { bar: 0, beat: 0.0 } }

    pub fn at_bar(bar: i32) -> Self { Self { bar, beat: 0.0 } }

    /// Convert to absolute beat position.
    pub fn to_absolute_beat(&self, ts: &TimeSignature) -> f32 {
        self.bar as f32 * ts.nb_natural_beats() + self.beat
    }

    /// Create from absolute beat position.
    pub fn from_absolute_beat(abs_beat: f32, ts: &TimeSignature) -> Self {
        let nb = ts.nb_natural_beats();
        let bar = (abs_beat / nb) as i32;
        let beat = abs_beat - bar as f32 * nb;
        Self { bar, beat }
    }

    pub fn beat_int(&self) -> i32 { self.beat as i32 }
    pub fn beat_frac(&self) -> f32 { self.beat - self.beat as i32 as f32 }
}
